File: packages/device_calendar_plus_android/android/src/main/kotlin/to/bullet/device_calendar_plus_android/RecurrenceAnchor.kt
package to.bullet.device_calendar_plus_android

import java.util.Calendar
import java.util.TimeZone

/**
 * Picks the anchor (DTSTART) for a series whose recurrence rule is being
 * replaced.
 *
 * RFC 5545 wants DTSTART to be a day the RRULE generates, and both Android's
 * Calendar Provider and iOS EventKit treat one that isn't as an extra first
 * occurrence — the orphaned Saturday of #140. [firstMatch] walks forward from
 * the intended anchor to the first day the rule generates, so the anchor of a
 * series switched to a new pattern lands on that pattern. iOS's counterpart is
 * `RecurrenceAnchor.swift`; the two must agree.
 *
 * Covers the RRULE subset the plugin models: FREQ, BYDAY (with ordinals),
 * BYMONTHDAY, BYMONTH, and BYSETPOS on monthly and yearly rules. INTERVAL and
 * WKST don't affect which days a rule can generate from a given anchor (the
 * anchor itself fixes the interval's phase), and a BYSETPOS on a daily or
 * weekly rule is outside the modelled subset, so all three are ignored.
 */
internal object RecurrenceAnchor {
    private enum class Freq { DAILY, WEEKLY, MONTHLY, YEARLY }

    /** One BYDAY entry: [weekday] is a `Calendar.DAY_OF_WEEK`, [ordinal] 0 = every. */
    private data class ByDay(val ordinal: Int, val weekday: Int)

    private data class Rule(
        val freq: Freq,
        val byDay: List<ByDay>,
        val byMonthDay: List<Int>,
        val byMonth: List<Int>,
        val bySetPos: List<Int>
    )

    private val weekdayCodes = mapOf(
        "SU" to Calendar.SUNDAY, "MO" to Calendar.MONDAY, "TU" to Calendar.TUESDAY,
        "WE" to Calendar.WEDNESDAY, "TH" to Calendar.THURSDAY, "FR" to Calendar.FRIDAY,
        "SA" to Calendar.SATURDAY
    )

    /** Five years: enough for a rule pinned to 29 February. */
    private const val MAX_LOOKAHEAD_DAYS = 5 * 366

    /**
     * The first instant on or after [fromMillis] whose calendar day (in [tz])
     * the rule generates, at [fromMillis]'s wall-clock time. Parts the rule
     * leaves implicit (no BYDAY, BYMONTHDAY or BYMONTH) come from [fromMillis]
     * itself, as they would from DTSTART — so a rule that already fits its
     * anchor returns [fromMillis] unchanged — as does a rule outside the
     * modelled subset (no FREQ this object knows), which is assumed to fit
     * its anchor. Null when the rule generates nothing within five years, so
     * the caller can refuse rather than anchor the series off-rule.
     */
    fun firstMatch(rrule: String, fromMillis: Long, tz: TimeZone): Long? {
        val cal = Calendar.getInstance(tz).apply { timeInMillis = fromMillis }
        val rule = parse(rrule)?.impliedBy(cal) ?: return fromMillis
        val matcher = Matcher(rule)
        repeat(MAX_LOOKAHEAD_DAYS) {
            if (matcher.generates(cal)) return cal.timeInMillis
            // Whole-day steps keep the wall-clock time across DST (mirrors
            // shiftDate).
            cal.add(Calendar.DAY_OF_MONTH, 1)
        }
        return null
    }

    private fun parse(rrule: String): Rule? {
        val params = RruleString.params(rrule)
        val freq = when (params["FREQ"]?.uppercase()) {
            "DAILY" -> Freq.DAILY
            "WEEKLY" -> Freq.WEEKLY
            "MONTHLY" -> Freq.MONTHLY
            "YEARLY" -> Freq.YEARLY
            else -> return null
        }
        fun ints(key: String): List<Int> =
            params[key]?.split(",")?.mapNotNull { it.trim().toIntOrNull() } ?: emptyList()
        val byDay = params["BYDAY"]?.split(",")
            ?.mapNotNull { parseByDay(it.trim().uppercase()) } ?: emptyList()
        return Rule(
            freq = freq,
            byDay = byDay,
            byMonthDay = ints("BYMONTHDAY"),
            byMonth = ints("BYMONTH"),
            bySetPos = ints("BYSETPOS")
        )
    }

    /** "TU", "2TU", "-1FR", "+1MO". */
    private fun parseByDay(value: String): ByDay? {
        if (value.length < 2) return null
        val weekday = weekdayCodes[value.takeLast(2)] ?: return null
        val prefix = value.dropLast(2)
        val ordinal = if (prefix.isEmpty()) 0 else (prefix.toIntOrNull() ?: return null)
        return ByDay(ordinal, weekday)
    }

    /**
     * Fills in the parts the rule leaves implicit from [anchor], as RFC 5545
     * fills them from DTSTART: a WEEKLY rule with no BYDAY runs on the
     * anchor's weekday, a MONTHLY rule with neither BYDAY nor BYMONTHDAY on
     * its day of the month, and a YEARLY rule likewise, in the anchor's month
     * unless BYMONTH lists its own. The one YEARLY form left alone is BYDAY on
     * its own, whose ordinals count within the year. A DAILY rule has nothing
     * implicit: each of its parts only filters.
     */
    private fun Rule.impliedBy(anchor: Calendar): Rule {
        val weekday = anchor.get(Calendar.DAY_OF_WEEK)
        val dayOfMonth = listOf(anchor.get(Calendar.DAY_OF_MONTH))
        val month = listOf(anchor.get(Calendar.MONTH) + 1)
        val ownDay = byDay.isEmpty() && byMonthDay.isEmpty()
        return when (freq) {
            Freq.DAILY -> this
            // A WEEKLY rule ignores BYMONTHDAY.
            Freq.WEEKLY -> copy(
                byDay = byDay.ifEmpty { listOf(ByDay(0, weekday)) },
                byMonthDay = emptyList()
            )
            Freq.MONTHLY -> if (ownDay) copy(byMonthDay = dayOfMonth) else this
            Freq.YEARLY ->
                if (byMonth.isEmpty() && byMonthDay.isEmpty() && byDay.isNotEmpty()) {
                    this
                } else {
                    copy(
                        byMonth = byMonth.ifEmpty { month },
                        byMonthDay = if (ownDay) dayOfMonth else byMonthDay
                    )
                }
        }
    }

    /**
     * Decides whether a day is in the rule's set for the period containing
     * it. A monthly or yearly period's set is built once and cached, since
     * [firstMatch] visits its days in order; the rule must already have had
     * its implicit parts filled in ([impliedBy]).
     */
    private class Matcher(private val rule: Rule) {
        private var cachedPeriod: Int? = null
        private var cachedDays: Set<Int> = emptySet()

        fun generates(day: Calendar): Boolean = when (rule.freq) {
            // Daily and weekly rules are decided a day at a time: whether a
            // day is in a weekly set doesn't depend on where the week starts.
            Freq.DAILY, Freq.WEEKLY -> dayPasses(day)
            Freq.MONTHLY, Freq.YEARLY -> {
                val period = periodKey(day)
                if (period != cachedPeriod) {
                    cachedPeriod = period
                    cachedDays = periodDays(day)
                }
                dayKey(day) in cachedDays
            }
        }

        /**
         * BYSETPOS keeps only the listed positions of a month's or year's set
         * (1-based; negative from the end).
         */
        private fun applySetPos(days: Set<Int>): Set<Int> {
            if (rule.bySetPos.isEmpty()) return days
            val ordered = days.sorted()
            return rule.bySetPos.mapNotNull { pos ->
                val index = if (pos > 0) pos - 1 else ordered.size + pos
                ordered.getOrNull(index)
            }.toSet()
        }

        private fun dayKey(c: Calendar): Int =
            c.get(Calendar.YEAR) * 1000 + c.get(Calendar.DAY_OF_YEAR)

        private fun periodKey(day: Calendar): Int = when (rule.freq) {
            Freq.MONTHLY -> day.get(Calendar.YEAR) * 100 + day.get(Calendar.MONTH)
            else -> day.get(Calendar.YEAR)
        }

        private fun periodDays(day: Calendar): Set<Int> = when (rule.freq) {
            Freq.MONTHLY -> if (inByMonth(day)) applySetPos(monthDays(day)) else emptySet()
            else -> applySetPos(yearDays(day))
        }

        /**
         * In a DAILY or WEEKLY rule every BYxxx part only filters (no
         * ordinals apply).
         */
        private fun dayPasses(day: Calendar): Boolean {
            if (!inByMonth(day)) return false
            val dom = day.get(Calendar.DAY_OF_MONTH)
            val length = day.getActualMaximum(Calendar.DAY_OF_MONTH)
            if (rule.byMonthDay.isNotEmpty() && !matchesMonthDay(dom, length)) return false
            val weekday = day.get(Calendar.DAY_OF_WEEK)
            return rule.byDay.isEmpty() || rule.byDay.any { it.weekday == weekday }
        }

        /** The days of [day]'s year the rule generates. */
        private fun yearDays(day: Calendar): Set<Int> {
            // BYDAY on its own (the one form impliedBy leaves without a
            // BYMONTH): ordinals count within the year ("the 20th Monday").
            // Otherwise they count within the month, and the yearly set is
            // the listed months' sets.
            if (rule.byMonth.isEmpty()) {
                val c = (day.clone() as Calendar).apply { set(Calendar.DAY_OF_YEAR, 1) }
                val yearLength = c.getActualMaximum(Calendar.DAY_OF_YEAR)
                val days = mutableSetOf<Int>()
                for (doy in 1..yearLength) {
                    c.set(Calendar.DAY_OF_YEAR, doy)
                    if (matchesByDay(c, nthInPeriod(doy, yearLength))) days += dayKey(c)
                }
                return days
            }
            val days = mutableSetOf<Int>()
            for (month in rule.byMonth) {
                if (month !in 1..12) continue
                val inMonth = (day.clone() as Calendar).apply {
                    set(Calendar.DAY_OF_MONTH, 1)
                    set(Calendar.MONTH, month - 1)
                }
                days += monthDays(inMonth)
            }
            return days
        }

        /** The days of [day]'s month the rule generates. */
        private fun monthDays(day: Calendar): Set<Int> {
            val first = (day.clone() as Calendar).apply { set(Calendar.DAY_OF_MONTH, 1) }
            val length = first.getActualMaximum(Calendar.DAY_OF_MONTH)
            val days = mutableSetOf<Int>()
            val c = first.clone() as Calendar
            for (dom in 1..length) {
                c.set(Calendar.DAY_OF_MONTH, dom)
                // Per RFC 5545, BYDAY limits a BYMONTHDAY set and expands
                // otherwise (impliedBy guarantees one of the two is present).
                val included = if (rule.byMonthDay.isNotEmpty()) {
                    matchesMonthDay(dom, length) &&
                        (rule.byDay.isEmpty() || matchesByDay(c, nthInPeriod(dom, length)))
                } else {
                    matchesByDay(c, nthInPeriod(dom, length))
                }
                if (included) days += dayKey(c)
            }
            return days
        }

        private fun matchesMonthDay(dom: Int, monthLength: Int): Boolean =
            rule.byMonthDay.any { it == dom || (it < 0 && monthLength + 1 + it == dom) }

        /** A day's weekday ordinal within its period, counted from the start and the end. */
        private data class Nth(val fromStart: Int, val fromEnd: Int)

        /**
         * The ordinal of the day at [position] (1-based) in a period of
         * [periodLength] days: which length is passed decides whether "2MO"
         * counts within the month or within the year.
         */
        private fun nthInPeriod(position: Int, periodLength: Int) =
            Nth((position - 1) / 7 + 1, -((periodLength - position) / 7 + 1))

        /** BYDAY membership: the weekday, and the ordinal when one is given. */
        private fun matchesByDay(c: Calendar, nth: Nth): Boolean {
            val weekday = c.get(Calendar.DAY_OF_WEEK)
            return rule.byDay.any {
                it.weekday == weekday &&
                    (it.ordinal == 0 || it.ordinal == nth.fromStart || it.ordinal == nth.fromEnd)
            }
        }

        private fun inByMonth(c: Calendar): Boolean =
            rule.byMonth.isEmpty() || (c.get(Calendar.MONTH) + 1) in rule.byMonth
    }
}

File: packages/device_calendar_plus_ios/ios/device_calendar_plus_ios/Sources/device_calendar_plus_ios/RecurrenceAnchor.swift
import EventKit
import Foundation

/// Picks the anchor (start) for a series whose recurrence rule is being
/// replaced.
///
/// RFC 5545 wants a series' start to be a day its rule generates, and both
/// iOS EventKit and Android's Calendar Provider treat one that isn't as an
/// extra first occurrence — the orphaned Saturday of #140. `firstMatch` walks
/// forward from the intended anchor to the first day the rule generates, so
/// the anchor of a series switched to a new pattern lands on that pattern.
/// Android's counterpart is `RecurrenceAnchor.kt`; the two must agree.
///
/// Covers the RRULE subset the plugin models: FREQ, BYDAY (with ordinals),
/// BYMONTHDAY, BYMONTH, and BYSETPOS on monthly and yearly rules. INTERVAL
/// and WKST don't affect which days a rule can generate from a given anchor
/// (the anchor itself fixes the interval's phase), and a BYSETPOS on a daily
/// or weekly rule is outside the modelled subset, so all three are ignored.
enum RecurrenceAnchor {
  /// Five years: enough for a rule pinned to 29 February.
  private static let maxLookaheadDays = 5 * 366

  /// The first instant on or after `from` whose calendar day (in `timeZone`)
  /// `rule` generates, at `from`'s wall-clock time. Parts the rule leaves
  /// implicit (no BYDAY, BYMONTHDAY or BYMONTH) come from `from` itself, as
  /// they would from the start date — so a rule that already fits its anchor
  /// returns `from` unchanged. Nil when the rule generates nothing within
  /// five years, so the caller can refuse rather than anchor the series
  /// off-rule.
  static func firstMatch(
    of rule: EKRecurrenceRule,
    onOrAfter from: Date,
    timeZone: TimeZone
  ) -> Date? {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    let matcher = Matcher(
      rule: Rule(rule).impliedBy(anchor: from, calendar: calendar), calendar: calendar
    )
    var day = from
    for _ in 0..<maxLookaheadDays {
      if matcher.generates(day) { return day }
      // Whole-day steps keep the wall-clock time across DST (mirrors
      // shiftStart).
      guard let next = calendar.date(byAdding: .day, value: 1, to: day) else { return nil }
      day = next
    }
    return nil
  }

  /// One BYDAY entry: `weekday` is a Gregorian weekday (1 = Sunday),
  /// `ordinal` 0 = every.
  private struct ByDay {
    let ordinal: Int
    let weekday: Int
  }

  /// A day's weekday ordinal within its period, counted from the start and
  /// the end.
  private struct Nth {
    let fromStart: Int
    let fromEnd: Int
  }

  /// The parts of an `EKRecurrenceRule` that decide which days it generates.
  private struct Rule {
    let frequency: EKRecurrenceFrequency
    var byDay: [ByDay]
    var byMonthDay: [Int]
    var byMonth: [Int]
    let bySetPos: [Int]

    /// The modelled parts of `rule`, as written.
    init(_ rule: EKRecurrenceRule) {
      frequency = rule.frequency
      byDay = (rule.daysOfTheWeek ?? []).map {
        ByDay(ordinal: $0.weekNumber, weekday: $0.dayOfTheWeek.rawValue)
      }
      byMonthDay = (rule.daysOfTheMonth ?? []).map { $0.intValue }
      byMonth = (rule.monthsOfTheYear ?? []).map { $0.intValue }
      bySetPos = (rule.setPositions ?? []).map { $0.intValue }
    }

    /// Fills in the parts the rule leaves implicit from `anchor`, as RFC
    /// 5545 fills them from DTSTART: a WEEKLY rule with no BYDAY runs on the
    /// anchor's weekday, a MONTHLY rule with neither BYDAY nor BYMONTHDAY on
    /// its day of the month, and a YEARLY rule likewise, in the anchor's
    /// month unless BYMONTH lists its own. The one YEARLY form left alone is
    /// BYDAY on its own, whose ordinals count within the year. A DAILY rule
    /// has nothing implicit: each of its parts only filters.
    func impliedBy(anchor: Date, calendar: Calendar) -> Rule {
      let weekday = calendar.component(.weekday, from: anchor)
      let dayOfMonth = [calendar.component(.day, from: anchor)]
      let month = [calendar.component(.month, from: anchor)]
      let ownDay = byDay.isEmpty && byMonthDay.isEmpty
      var implied = self
      switch frequency {
      case .daily:
        break
      case .weekly:
        if byDay.isEmpty { implied.byDay = [ByDay(ordinal: 0, weekday: weekday)] }
        // A WEEKLY rule ignores BYMONTHDAY.
        implied.byMonthDay = []
      case .monthly:
        if ownDay { implied.byMonthDay = dayOfMonth }
      case .yearly:
        if byMonth.isEmpty && byMonthDay.isEmpty && !byDay.isEmpty { break }
        if byMonth.isEmpty { implied.byMonth = month }
        if ownDay { implied.byMonthDay = dayOfMonth }
      @unknown default:
        break
      }
      return implied
    }
  }

  /// Decides whether a day is in the rule's set for the period containing
  /// it. A monthly or yearly period's set is built once and cached, since
  /// `firstMatch` visits its days in order; the rule must already have had
  /// its implicit parts filled in (`Rule.impliedBy`).
  private final class Matcher {
    private let calendar: Calendar
    private let frequency: EKRecurrenceFrequency
    private let byDay: [ByDay]
    private let byMonthDay: [Int]
    private let byMonth: [Int]
    private let bySetPos: [Int]
    private var cachedPeriod: Int?
    private var cachedDays: Set<Int> = []

    init(rule: Rule, calendar: Calendar) {
      self.calendar = calendar
      frequency = rule.frequency
      byDay = rule.byDay
      byMonthDay = rule.byMonthDay
      byMonth = rule.byMonth
      bySetPos = rule.bySetPos
    }

    func generates(_ day: Date) -> Bool {
      switch frequency {
      // Daily and weekly rules are decided a day at a time: whether a day is
      // in a weekly set doesn't depend on where the week starts.
      case .daily, .weekly:
        return dayPasses(day)
      case .monthly, .yearly:
        let period = periodKey(day)
        if period != cachedPeriod {
          cachedPeriod = period
          cachedDays = periodDays(day)
        }
        return cachedDays.contains(dayKey(day))
      @unknown default:
        return false
      }
    }

    /// BYSETPOS keeps only the listed positions of a month's or year's set
    /// (1-based; negative from the end).
    private func applySetPos(_ days: Set<Int>) -> Set<Int> {
      if bySetPos.isEmpty { return days }
      let ordered = days.sorted()
      return Set(bySetPos.compactMap { pos -> Int? in
        let index = pos > 0 ? pos - 1 : ordered.count + pos
        return ordered.indices.contains(index) ? ordered[index] : nil
      })
    }

    private func dayKey(_ day: Date) -> Int {
      let year = calendar.component(.year, from: day)
      let dayOfYear = calendar.ordinality(of: .day, in: .year, for: day) ?? 0
      return year * 1000 + dayOfYear
    }

    private func periodKey(_ day: Date) -> Int {
      if frequency == .monthly {
        return calendar.component(.year, from: day) * 100 + calendar.component(.month, from: day)
      }
      return calendar.component(.year, from: day)
    }

    private func periodDays(_ day: Date) -> Set<Int> {
      if frequency == .monthly {
        return inByMonth(day) ? applySetPos(monthDays(day)) : []
      }
      return applySetPos(yearDays(day))
    }

    /// In a DAILY or WEEKLY rule every BYxxx part only filters (no ordinals
    /// apply).
    private func dayPasses(_ day: Date) -> Bool {
      guard inByMonth(day) else { return false }
      let dom = calendar.component(.day, from: day)
      let length = monthLength(of: day)
      if !byMonthDay.isEmpty && !matchesMonthDay(dom, monthLength: length) { return false }
      let weekday = calendar.component(.weekday, from: day)
      return byDay.isEmpty || byDay.contains { $0.weekday == weekday }
    }

    /// The days of `day`'s year the rule generates.
    private func yearDays(_ day: Date) -> Set<Int> {
      let year = calendar.component(.year, from: day)
      // BYDAY on its own (the one form `impliedBy` leaves without a
      // BYMONTH): ordinals count within the year ("the 20th Monday").
      // Otherwise they count within the month, and the yearly set is the
      // listed months' sets.
      if byMonth.isEmpty {
        guard let first = date(year: year, month: 1, day: 1) else { return [] }
        let yearLength = calendar.range(of: .day, in: .year, for: first)?.count ?? 365
        var days = Set<Int>()
        for doy in 1...yearLength {
          guard let current = calendar.date(byAdding: .day, value: doy - 1, to: first) else { break }
          if matchesByDay(current, nth: nthInPeriod(doy, periodLength: yearLength)) {
            days.insert(dayKey(current))
          }
        }
        return days
      }
      var days = Set<Int>()
      for month in byMonth where (1...12).contains(month) {
        guard let inMonth = date(year: year, month: month, day: 1) else { continue }
        days.formUnion(monthDays(inMonth))
      }
      return days
    }

    /// The days of `day`'s month the rule generates.
    private func monthDays(_ day: Date) -> Set<Int> {
      let year = calendar.component(.year, from: day)
      let month = calendar.component(.month, from: day)
      let length = monthLength(of: day)
      var days = Set<Int>()
      for dom in 1...length {
        guard let current = date(year: year, month: month, day: dom) else { continue }
        let nth = nthInPeriod(dom, periodLength: length)
        // Per RFC 5545, BYDAY limits a BYMONTHDAY set and expands otherwise
        // (`impliedBy` guarantees one of the two is present).
        let included: Bool
        if !byMonthDay.isEmpty {
          included = matchesMonthDay(dom, monthLength: length)
            && (byDay.isEmpty || matchesByDay(current, nth: nth))
        } else {
          included = matchesByDay(current, nth: nth)
        }
        if included { days.insert(dayKey(current)) }
      }
      return days
    }

    /// The ordinal of the day at `position` (1-based) in a period of
    /// `periodLength` days: which length is passed decides whether "2MO"
    /// counts within the month or within the year.
    private func nthInPeriod(_ position: Int, periodLength: Int) -> Nth {
      return Nth(fromStart: (position - 1) / 7 + 1, fromEnd: -((periodLength - position) / 7 + 1))
    }

    private func matchesMonthDay(_ dom: Int, monthLength: Int) -> Bool {
      return byMonthDay.contains { $0 == dom || ($0 < 0 && monthLength + 1 + $0 == dom) }
    }

    /// BYDAY membership: the weekday, and the ordinal when one is given.
    private func matchesByDay(_ day: Date, nth: Nth) -> Bool {
      let weekday = calendar.component(.weekday, from: day)
      return byDay.contains {
        $0.weekday == weekday
          && ($0.ordinal == 0 || $0.ordinal == nth.fromStart || $0.ordinal == nth.fromEnd)
      }
    }

    private func inByMonth(_ day: Date) -> Bool {
      return byMonth.isEmpty || byMonth.contains(calendar.component(.month, from: day))
    }

    private func monthLength(of day: Date) -> Int {
      return calendar.range(of: .day, in: .month, for: day)?.count ?? 31
    }

    /// Noon on the given date: only the day matters for set membership, and
    /// noon sidesteps DST transitions at midnight.
    private func date(year: Int, month: Int, day: Int) -> Date? {
      return calendar.date(from: DateComponents(year: year, month: month, day: day, hour: 12))
    }
  }
}
